roadmap/implementers-guide/src/node/utility/availability-store.md

```diff
@@ -65,104 +59,127 @@ digraph {
 }
 ```
 
-## Protocol
+## Database Schema.
 
-Input: [`AvailabilityStoreMessage`][ASM]
+We use an underlying Key-Value database where we assume we have the following operations available:
+  * `write(key, value)`
+  * `read(key) -> Option<value>`
+  * `iter_by_prefix(prefix) -> Iterator<(key, value)>` - gives all keys and values in lexicographical order starting from `prefix`
 
-Output:
-- [`RuntimeApiMessage`][RAM]
+We use this database to encode the following schema:
 
-## Functionality
+```
+"LastFinalized" -> BlockNumber
+("available", CandidateHash) -> Option<AvailableData>
+("chunk", CandidateHash, u32) -> Option<ErasureChunk>
+("meta", CandidateHash) -> Option<CandidateMeta>
 
-On `ActiveLeavesUpdate`:
+("appears_in_unfinalized", BlockNumber, BlockHash, CandidateHash) -> Option<()>
+("prune_by_time", Timestamp, CandidateHash) -> Option<()>
+```
 
-For each head in the `activated` list:
-	- Note any new candidates backed in the block. Update pruning records for any stored `PoVBlock`s.
-	- Note any newly-included candidates backed in the block. Update pruning records for any stored availability chunks.
+The meta information that we track per-candidate is defined as the `CandidateMeta` struct
 
-On `OverseerSignal::BlockFinalized(_)` events:
+```rust
+struct CandidateMeta {
+  state: State,
+  data_available: bool,
+  chunks_stored: Bitfield,
+}
 
-- Handle all pruning based on the newly-finalized block.
+enum State {
+  /// Candidate data was first observed at the given time but is not available in any block.
+  Unavailable(Timestamp),
+  /// The candidate was first observed at the given time and was included in the given list of unfinalized blocks, which may be
+  /// empty. The timestamp here is not used for pruning.
+  Unfinalized(Timestamp, Vec<(BlockNumber, BlockHash)>),
+  /// Candidate data has appeared in a finalized block and did so at the given time.
+  Finalized(Timestamp)
+}
+```
 
-On `QueryPoV` message:
+We maintain the invariant that if a candidate has a meta entry, its available data exists on disk if `data_available` is true. All chunks mentioned in the meta entry are available.
```

Comment:
Very good point.  We might've an only partially reconstructed block in which case we should keep all those pieces and eventually add some paths for sorting out the mess with our peers.  I donno how much of that exists.